File: python/aidoodle/ai/mcts.py
from dataclasses import dataclass, field
import enum
import math
import random
from typing import List, Optional, Union, Iterable, TypeVar, Dict

from aidoodle.core import Engine, Game, Move, Player


C = math.sqrt(2)  # from literature
EPS = 1e-12  # for numerical stability
VERBOSE = 0
MAX_DEPTH = 10000

Numeric = Union[float, int]
T = TypeVar('T')


@dataclass
class Edge:
    move: Move
    w: float = 0.0
    s: int = 0

    def __repr__(self) -> str:
        return f"Edge({self.move}, w={self.w}, s={self.s})"


@dataclass
class Node:
    game: Game
    edges: List[Edge] = field(default_factory=list)

    def __repr__(self) -> str:
        g = str(hash(self.game) % 1000) + '..'
        return f"Node(n_children={len(self.edges)}, game={g})"

    def __hash__(self) -> int:
        return hash(self.game)


_Players = List[Player]
_Edges = List[Edge]
_Nodes = List[Node]
Cache = Dict[Game, Node]
MaybeNode = Optional[Node]


class Strategy(enum.Enum):
    random = 0
    ucb1 = 1


def _selectmax(keys: Iterable[T], vals: Iterable[Numeric]) -> T:
    return max(zip(keys, vals), key=lambda tup: tup[1])[0]


def _ucb1_values(edges: _Edges, c: float = C) -> List[float]:
    s_tot = sum(edge.s for edge in edges)
    const = c * math.log(s_tot + 1)
    vals = [e.w / (e.s + EPS) + const / math.sqrt(e.s + EPS) for e in edges]
    return vals


def select_ucb1(edges: _Edges, c: float = C) -> Edge:
    vals = _ucb1_values(edges=edges, c=c)
    edge = _selectmax(edges, vals)
    return edge


def select(edges: _Edges, strategy: Strategy = Strategy.ucb1) -> Edge:
    if strategy == Strategy.random:
        return random.choice(edges)
    if strategy == Strategy.ucb1:
        return select_ucb1(edges)
    raise ValueError("Unknown strategy")


def choose_edge(edges: _Edges) -> Edge:
    edge = _selectmax(edges, (e.s for e in edges))
    if VERBOSE:
        print(f"Number of visits: {edge.s}, wins: {100*edge.w/edge.s:.1f}%")
    return edge


def expand(node: Node, engine: Engine) -> None:
    # Careful: if a move is the identity move, there will be an
    # infinite recursion
    moves: List[Move] = engine.get_legal_moves(node.game)
    edges = [Edge(move) for move in moves]
    assert not node.edges
    node.edges = edges


def simulate(game: Game, engine: Engine) -> float:
    # init a game with random players
    game = engine.init_game(
        board=game.board,
        player_idx=game.player_idx,
    )

    if VERBOSE:
        print("-" * 40)
        print(game.board)

    while not game.winner:
        # by default uses random play
        move = random.choice(engine.get_legal_moves(game))
        game = engine.make_move(game=game, move=move)

    if VERBOSE:
        print(game.board, end=' ')
        print(game.winner)

    score: float = engine.game_score(game)
    return score


def _update_edge(edge: Edge, value: float) -> None:
    edge.s += 1
    edge.w += value



COUNTER = 0
def update(edges: _Edges, players: _Players, value: float) -> None:
    value_other = 1 - value
    for edge, player in zip(edges, players):
        if player == 1:
            _update_edge(edge=edge, value=value)
        elif player == 2:
            _update_edge(edge=edge, value=value_other)


def _retrieve_node(game: Game, cache: Cache) -> Node:
    maybe_node: MaybeNode = cache.get(game)
    if maybe_node is not None:
        return maybe_node

    node = Node(game=game)
    cache[game] = node
    return node


def search_iteration(
        node: Node,
        engine: Engine,
        cache: Cache,
        strategy: Strategy = Strategy.ucb1,
) -> None:
    cache[node.game] = node
    edges: _Edges = []
    players: _Players = []

    # selection
    while node.edges:
        edge = select(node.edges, strategy=strategy)
        edges.append(edge)
        players.append(node.game.player)
        game = engine.make_move(game=node.game, move=edge.move)
        node = _retrieve_node(game=game, cache=cache)  # updates cache if necessary

        if len(edges) > MAX_DEPTH:
            raise RuntimeError(f"Max depth of {MAX_DEPTH} in tree search encountered, "
                               "are there cycles in the game tree?")

    # expansion
    expand(node, engine=engine)

    if node.edges:  # game end not reached
        # -> choose random move
        edge = random.choice(node.edges)
        game = engine.make_move(game=node.game, move=edge.move)
        edges.append(edge)
        players.append(game.player)
    else:  # end state reached
        game = node.game

    # simulate
    value = simulate(game, engine=engine)

    # update
    update(edges, players=players, value=value)

File: typescript/src/mcts.ts
import { BaseEngine, Move, Game, Player, gameToString } from "./games/tictactoe";
import { randChoice, sum } from './utils';

const C = Math.SQRT2;
const EPS = 1e-12;
const MAX_DEPTH = 10000;
const N_ITER = 100;

interface Edge {
    move: Move;
    w: number;
    s: number;
};

interface Node {
    game: Game;
    edges: Edges;
};

interface Cache {
    [gameString: string]: Node;
};

type Players = Player[];
type Edges = Edge[];


const selectmax = (edges: Edges, vals: number[]): Edge => {
    if (vals.length < 1) {
        throw new Error("Cannot select from empty array");
    };
    if (vals.length !== edges.length) {
        throw new Error("Keys and values must be of equal length");
    }

    const argmax = vals.map(
        (x, i) => [x, i]
    ).reduce((r, a) => (a[0] > r[0] ? a : r))[1];
    return edges[argmax];
};

const ucb1Values = (edges: Edges): number[] => {
    const sVals = edges.map((edge) => edge.s);
    const sTot = sum(sVals);
    const c = C * Math.log(sTot + 1);
    const vals = edges.map((e) => e.w / (e.s + EPS) + c / Math.sqrt(e.s + EPS));
    return vals;
};

const selectUcb1 = (edges: Edges): Edge => {
    const vals = ucb1Values(edges);
    const edge = selectmax(edges, vals);
    return edge;
};

const expand = (node: Node, engine: BaseEngine) => {
    // Careful: if a move is the identity move, there will be an infinte loop
    if (node.edges.length > 0) {
        throw new Error("Node shouldn't have any edges");
    };

    const moves = engine.getLegalMoves(node.game);
    let edges: Edges = [];
    moves.forEach((move) => edges.push({ move: move, w: 0, s: 0 }));

    node.edges = edges;
};

const simulate = (game: Game, engine: BaseEngine): number => {
    var newGame = engine.initGame(game.board, game.playerIdx);

    let winner = engine.determineWinner(newGame);
    while (winner === undefined) {
        var move = randChoice(engine.getLegalMoves(newGame));
        var newGame = engine.makeMove(newGame, move);
        console.log(game);
        console.log(move);
    };

    const score = engine.scoreGame(newGame);
    return score;
};

const updateEdge = (edge: Edge, value: number) => {
    edge.s = edge.s + 1;
    edge.w = edge.w + value;
};

const update = (edges: Edges, players: Players, value: number) => {
    if (edges.length !== players.length) {
        throw new Error("Number of edges and number of players not equal");
    };

    const valueOther = 1 - value;
    for (let i = 0; edges.length; i++) {
        var edge = edges[i];
        var player = players[i];
        (player === 1) ? updateEdge(edge, value) : updateEdge(edge, valueOther);
    };
};

const retrieveNode = (game: Game, cache: Cache): Node => {
    const maybeNode: Node | undefined = cache[gameToString(game)];
    if (maybeNode !== undefined) {
        return maybeNode;
    };

    const node: Node = { game: game, edges: [] };
    cache[gameToString(game)] = node;
    return node;
}

export const search = (node: Node, engine: BaseEngine, cache: Cache) => {
    cache[gameToString(node.game)] = node;
    var edges: Edges = [];
    var players: Players = [];

    while (node.edges.length > 0) {
        const edge = selectUcb1(node.edges);
        edges.push(edge);
        const player = node.game.players[node.game.playerIdx];
        players.push(player);

        const game = engine.makeMove(node.game, edge.move);
        node = retrieveNode(game, cache);

        if (edges.length > MAX_DEPTH) {
            throw new Error("Max search depth has been exceeded, infinite loop?");
        };
    };

    expand(node, engine);
    if (node.edges.length > 0) {  // game has not ended
        const edge = randChoice(node.edges);
        var game = engine.makeMove(node.game, edge.move);
        edges.push(edge);
        players.push(game.players[game.playerIdx]);
        const value = simulate(game, engine);
        update(edges, players, value);
    } else {
        const game = node.game;
        const value = simulate(game, engine);
        update(edges, players, value);
    };
};

export const mctsMove = (game: Game, engine: BaseEngine): Move => {
    const edges: Edges = [];
    const root: Node = { game: game, edges: edges };
    const cache: Cache = {};

    for (let _ = 0; N_ITER; _++) {
        search(root, engine, cache);
    };

    const values = edges.map((edge) => edge.s)
    const edge = selectmax(root.edges, values);
    return edge.move;
}